README.md
scrapli_asyncssh
================

This is the asyncssh plugin for [scrapli](https://github.com/carlmontanari/scrapli).

If you would like to use scrapli, but prefer to use asyncssh as a transport so you can do super fast and awesome
 asyncio stuff, then this is the library for you!
 
Please see the main [scrapli repo](https://github.com/carlmontanari/scrapli) for much more information.


CHANGELOG.md
CHANGELOG
=======

# XXXX.XX.XX
- Catch `BrokenPipeError` during connection close... going to assume that this is occurring due to connection 
  *already* being closed and given that that is what we are hoping to achieve in the `close` method that seems 
  alright with me. (note that this seemed to only crop up in github actions w/ ubuntu-latest and py3.8)


# 2020.12.23
- Catch `ConnectionLost` exception during read operations so we can raise a scrapli exception instead of the asyncssh
 exception
- Explicitly pass ssh config file to asyncssh -- in this way if the user sets ssh_config_file=False we will make 
  sure that asyncssh ignores the config file
- Bumped up the default timeout values as they were probably a little on the aggressive side -- its user 
  configurable anyway so folks can set this to whatever they feel like!


# 2020.10.10
- Minor internal updates to appease updated pylint/isort
- Add 3.9 to actions, update pins, try to behave like an adult :D
- Set preferred auth options at each authentication method to only try that explicit auth method
- Remove keepalive stuff for now (in line w/ scrapli core)
- Remove transport session locks


# 2020.07.04
- Disable ssh agent for now -- this will probably get supported/added in the future, but can cause hard to
 troubleshoot delays for now!
- Made transport timeout actually be used :)


# 2020.06.06
- First "real" release of scrapli_asyncssh -- still very early, but this has been working great in testing! Give it a
 shot!


tests/unit/transport/test_asyncssh_.py
from pathlib import Path

import pytest

import scrapli_asyncssh
from scrapli.exceptions import KeyVerificationFailed
from scrapli_asyncssh.transport import Transport

TEST_DATA_DIR = f"{Path(scrapli_asyncssh.__file__).parents[1]}/tests/test_data"


class DummyClass:
    pass


def test_creation():
    conn = Transport("localhost")
    assert conn.host == "localhost"
    assert conn.port == 22


@pytest.mark.parametrize(
    "test_host",
    [
        (
            "1.2.3.4",
            ["carl", "~/.ssh/mysshkey", 1234],
        ),
        (
            "5.6.7.8",
            ["somebodyelse", "~/.ssh/lastresortkey", 22],
        ),
        (
            "scrapli",
            ["scrapli", "~/.ssh/lastresortkey", 22],
        ),
    ],
    ids=["host_1.2.3.4", "catch_all", "specific_user_catch_all_key"],
)
def test__process_ssh_config(test_host):
    host = test_host[0]
    expected_auth_username = test_host[1][0]
    expected_private_key = test_host[1][1]
    expected_port = test_host[1][2]

    conn = Transport(host, ssh_config_file=f"{TEST_DATA_DIR}/files/_ssh_config")
    assert conn.host == host
    assert conn.auth_username == expected_auth_username
    assert conn.auth_private_key == str(Path(expected_private_key).expanduser())
    assert conn.port == expected_port


def test__verify_key_valid():
    conn = Transport("172.18.0.11")
    conn.ssh_known_hosts_file = f"{TEST_DATA_DIR}/files/_ssh_known_hosts"

    conn.session = DummyClass()

    def mock_export_public_key():
        return (
            b"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC+9q0c7+tuKT0+xS5JqMhlSoZ5gMuePUwMj1ELoij2vjoPj1Vk/+MvubDTr"
            b"/VGn6FwomQS9Ge3jNswk1mJN0SIcJuthg3OBN5LsQ/zEbh4RgrDnxaBjYkypabkTtOL3xTTd1mZBsa7+OvfGEb"
            b"+/qfv53wNT7Oy6K7fLhxaSm5bd5CioIV5i9SyOpzxy7ss2wPKX6pGaRx8GERfyfF2FnqyM/rLAYdiKHuuyJPwjFDxe2dRbOzpqmH"
            b"+RDd9lvggKaVzaL0XooXAhpDpz7BdD5efefwq6TysdLGtRvXEH0V/YhqodOCqntcjXTpRPX+Mi3fa8VS9FMS4qY5YKiLvRcil\n "
        )

    def mock_get_server_host_key():
        remote_server_key = DummyClass()
        remote_server_key.export_public_key = mock_export_public_key
        return remote_server_key

    conn.session.get_server_host_key = mock_get_server_host_key

    conn._verify_key_value()


def test__verify_key_invalid():
    conn = Transport("172.18.0.11")
    conn.ssh_known_hosts_file = f"{TEST_DATA_DIR}/files/_ssh_known_hosts"

    conn.session = DummyClass()

    def mock_export_public_key():
        return b"ssh-rsa blah\n "

    def mock_get_server_host_key():
        remote_server_key = DummyClass()
        remote_server_key.export_public_key = mock_export_public_key
        return remote_server_key

    conn.session.get_server_host_key = mock_get_server_host_key

    with pytest.raises(KeyVerificationFailed) as exc:
        conn._verify_key_value()

    assert str(exc.value) == "172.18.0.11 in known_hosts but public key does not match!"


def test__verify_key_not_found():
    conn = Transport("1.1.1.1")
    conn.ssh_known_hosts_file = f"{TEST_DATA_DIR}/files/_ssh_known_hosts"

    conn.session = DummyClass()

    def mock_export_public_key():
        return b"ssh-rsa blah\n "

    def mock_get_server_host_key():
        remote_server_key = DummyClass()
        remote_server_key.export_public_key = mock_export_public_key
        return remote_server_key

    conn.session.get_server_host_key = mock_get_server_host_key

    with pytest.raises(KeyVerificationFailed) as exc:
        conn._verify_key()

    assert str(exc.value) == "1.1.1.1 not in known_hosts!"


@pytest.mark.asyncio
async def test_open_verify_key():
    conn = Transport("172.18.0.11", auth_strict_key=True)
    conn.ssh_known_hosts_file = f"{TEST_DATA_DIR}/files/_ssh_known_hosts"

    conn.session = DummyClass()

    def mock_export_public_key():
        return (
            b"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC+9q0c7+tuKT0+xS5JqMhlSoZ5gMuePUwMj1ELoij2vjoPj1Vk/+MvubDTr"
            b"/VGn6FwomQS9Ge3jNswk1mJN0SIcJuthg3OBN5LsQ/zEbh4RgrDnxaBjYkypabkTtOL3xTTd1mZBsa7+OvfGEb"
            b"+/qfv53wNT7Oy6K7fLhxaSm5bd5CioIV5i9SyOpzxy7ss2wPKX6pGaRx8GERfyfF2FnqyM/rLAYdiKHuuyJPwjFDxe2dRbOzpqmH"
            b"+RDd9lvggKaVzaL0XooXAhpDpz7BdD5efefwq6TysdLGtRvXEH0V/YhqodOCqntcjXTpRPX+Mi3fa8VS9FMS4qY5YKiLvRcil\n "
        )

    def mock_get_server_host_key():
        remote_server_key = DummyClass()
        remote_server_key.export_public_key = mock_export_public_key
        return remote_server_key

    async def mock_authenticate():
        return True

    async def mock_open_session(**kwargs):
        return 1, 2, 3

    conn._authenticate = mock_authenticate
    conn.session.get_server_host_key = mock_get_server_host_key
    conn.session.open_session = mock_open_session

    await conn.open()


def test_set_timeout():
    conn = Transport("172.18.0.11")
    assert conn.timeout_transport == 30
    conn.set_timeout(999)
    assert conn.timeout_transport == 999


noxfile.py
"""scrapli_asyncssh.noxfile"""
import re
from pathlib import Path
from typing import Dict, List

import nox

nox.options.error_on_missing_interpreters = False
nox.options.stop_on_first_error = False

DEV_REQUIREMENTS: Dict[str, str] = {}

# this wouldn't work for other projects probably as its kinda hacky, but works just fine for scrapli
with open("requirements-dev.txt") as f:
    req_lines = f.readlines()
    dev_requirements_lines: List[str] = [
        line
        for line in req_lines
        if not line.startswith("-r") and not line.startswith("#") and not line.startswith("-e")
    ]
    dev_editable_requirements_lines: List[str] = [
        line for line in req_lines if line.startswith("-e")
    ]

for requirement in dev_requirements_lines:
    parsed_requirement = re.match(
        pattern=r"^([a-z0-9\-]+)([><=]{1,2}\S*)(?:.*)$", string=requirement, flags=re.I | re.M
    )
    DEV_REQUIREMENTS[parsed_requirement.groups()[0]] = parsed_requirement.groups()[1]

for requirement in dev_editable_requirements_lines:
    parsed_requirement = re.match(
        pattern=r"^-e\s.*(?:#egg=)(\w+)$", string=requirement, flags=re.I | re.M
    )
    DEV_REQUIREMENTS[parsed_requirement.groups()[0]] = requirement


@nox.session(python=["3.6", "3.7", "3.8", "3.9"])
def unit_tests(session):
    """
    Nox run unit tests

    Args:
        session: nox session

    Returns:
        N/A  # noqa: DAR202

    Raises:
        N/A

    """
    session.install("-r", "requirements-dev.txt")
    session.run(
        "pytest",
        "--cov=scrapli_asyncssh",
        "--cov-report",
        "html",
        "--cov-report",
        "term",
        "tests/unit",
        "-v",
    )


@nox.session(python=["3.9"])
def isort(session):
    """
    Nox run isort

    Args:
        session: nox session

    Returns:
        N/A  # noqa: DAR202

    Raises:
        N/A

    """
    session.install(f"isort{DEV_REQUIREMENTS['isort']}")
    session.run("isort", "-c", ".")


@nox.session(python=["3.9"])
def black(session):
    """
    Nox run black

    Args:
        session: nox session

    Returns:
        N/A  # noqa: DAR202

    Raises:
        N/A

    """
    session.install(f"black{DEV_REQUIREMENTS['black']}")
    session.run("black", "--check", ".")


@nox.session(python=["3.9"])
def pylama(session):
    """
    Nox run pylama

    Args:
        session: nox session

    Returns:
        N/A  # noqa: DAR202

    Raises:
        N/A

    """
    session.install("-r", "requirements-dev.txt")
    session.install("-e", ".")
    session.run("pylama", ".")


@nox.session(python=["3.9"])
def pydocstyle(session):
    """
    Nox run pydocstyle

    Args:
        session: nox session

    Returns:
        N/A  # noqa: DAR202

    Raises:
        N/A

    """
    session.install(f"pydocstyle{DEV_REQUIREMENTS['pydocstyle']}")
    session.run("pydocstyle", ".")


@nox.session(python=["3.9"])
def mypy(session):
    """
    Nox run mypy

    Args:
        session: nox session

    Returns:
        N/A  # noqa: DAR202

    Raises:
        N/A

    """
    session.install(f"mypy{DEV_REQUIREMENTS['mypy']}")
    session.install("-e", DEV_REQUIREMENTS["scrapli_stubs"].split()[1])
    session.env["MYPYPATH"] = f"{session.virtualenv.location}/src/scrapli-stubs"
    session.run("mypy", "--strict", "scrapli_asyncssh/")


@nox.session(python=["3.9"])
def darglint(session):
    """
    Nox run darglint

    Args:
        session: nox session

    Returns:
        N/A  # noqa: DAR202

    Raises:
        N/A

    """
    session.install(f"darglint{DEV_REQUIREMENTS['darglint']}")
    files_to_darglint = Path("scrapli_asyncssh").rglob("*.py")
    for file in files_to_darglint:
        session.run("darglint", f"{file.absolute()}")


scrapli_asyncssh/transport/asyncssh_.py
"""scrapli_asyncssh.transport.asyncssh_"""
import asyncio
from typing import Any, Dict, Optional, Tuple

from asyncssh import connect
from asyncssh.connection import SSHClientConnection
from asyncssh.misc import ConnectionLost, PermissionDenied
from asyncssh.stream import SSHReader, SSHWriter

from scrapli.decorators import requires_open_session
from scrapli.exceptions import KeyVerificationFailed, ScrapliAuthenticationFailed, ScrapliTimeout
from scrapli.ssh_config import SSHConfig, SSHKnownHosts
from scrapli.transport import AsyncTransport

ASYNCSSH_TRANSPORT_ARGS = (
    "auth_username",
    "auth_private_key",
    "auth_password",
    "auth_strict_key",
    "ssh_config_file",
    "ssh_known_hosts_file",
    "timeout_socket",
)


class AsyncSSHTransport(AsyncTransport):
    def __init__(
        self,
        host: str,
        port: int = 22,
        auth_username: str = "",
        auth_private_key: str = "",
        auth_password: str = "",
        auth_strict_key: bool = True,
        timeout_socket: int = 10,
        timeout_transport: int = 30,
        timeout_exit: bool = True,
        ssh_config_file: str = "",
        ssh_known_hosts_file: str = "",
    ) -> None:
        """
        AsyncSSHTransport Object

        AsyncSSHTransport <- AsyncTransport <- Transport (ABC)

        Args:
            host: host ip/name to connect to
            port: port to connect to
            auth_username: username for authentication
            auth_private_key: path to private key for authentication
            auth_password: password for authentication
            auth_strict_key: True/False to enforce strict key checking (default is True)
            timeout_socket: timeout for establishing socket in seconds
            timeout_transport: timeout for ssh transport in seconds
            timeout_exit: True/False close transport if timeout encountered
            ssh_config_file: string to path for ssh config file
            ssh_known_hosts_file: string to path for ssh known hosts file

        Returns:
            N/A  # noqa: DAR202

        Raises:
            N/A

        """
        cfg_port, cfg_user, cfg_private_key = self._process_ssh_config(host, ssh_config_file)
        port = cfg_port or port

        super().__init__(
            host,
            port,
            timeout_socket,
            timeout_transport,
            timeout_exit,
        )

        self.timeout_transport: int

        self.auth_username: str = auth_username or cfg_user
        self.auth_private_key: str = auth_private_key or cfg_private_key
        self.auth_password: str = auth_password
        self.auth_strict_key: bool = auth_strict_key
        self.ssh_config_file: str = ssh_config_file
        self.ssh_known_hosts_file: str = ssh_known_hosts_file

        self.session: SSHClientConnection
        self.stdout: SSHReader
        self.stdin: SSHWriter
        self.stderr: SSHReader

    @staticmethod
    def _process_ssh_config(host: str, ssh_config_file: str) -> Tuple[Optional[int], str, str]:
        """
        Method to parse ssh config file

        In the future this may move to be a 'helper' function as it should be very similar between
        asyncssh and and paramiko/ssh2-python... for now it can be a static method as there may be
        varying supported args between the transport drivers.

        Args:
            host: host to lookup in ssh config file
            ssh_config_file: string path to ssh config file; passed down from `Scrape`, or the
                `NetworkDriver` or subclasses of it, in most cases.

        Returns:
            Tuple: port to use for ssh, username to use for ssh, identity file (private key) to
                use for ssh auth

        Raises:
            N/A

        """
        ssh = SSHConfig(ssh_config_file)
        host_config = ssh.lookup(host)
        return host_config.port, host_config.user or "", host_config.identity_file or ""

    def _verify_key(self) -> None:
        """
        Verify target host public key, raise exception if invalid/unknown

        Args:
            N/A

        Returns:
            N/A  # noqa: DAR202

        Raises:
            KeyVerificationFailed: if host is not in known hosts

        """
        known_hosts = SSHKnownHosts(self.ssh_known_hosts_file)

        if self.host not in known_hosts.hosts.keys():
            raise KeyVerificationFailed(f"{self.host} not in known_hosts!")

    def _verify_key_value(self) -> None:
        """
        Verify target host public key, raise exception if invalid/unknown

        Args:
            N/A

        Returns:
            N/A  # noqa: DAR202

        Raises:
            KeyVerificationFailed: if host is in known hosts but public key does not match

        """
        known_hosts = SSHKnownHosts(self.ssh_known_hosts_file)

        remote_server_key = self.session.get_server_host_key()
        remote_public_key = remote_server_key.export_public_key().split()[1].decode()

        if known_hosts.hosts[self.host]["public_key"] != remote_public_key:
            raise KeyVerificationFailed(
                f"{self.host} in known_hosts but public key does not match!"
            )

    async def open(self) -> None:
        """
        Parent method to open session, authenticate and acquire shell

        Args:
            N/A

        Returns:
            N/A  # noqa: DAR202

        Raises:
            N/A

        """
        if self.auth_strict_key:
            self.logger.debug(f"Attempting to validate {self.host} public key is in known hosts")
            self._verify_key()

        await self._authenticate()

        if self.auth_strict_key:
            self.logger.debug(
                f"Attempting to validate {self.host} public key is in known hosts and is valid"
            )
            self._verify_key_value()

        # it seems we must pass a terminal type to force a pty(?) which i think we want in like...
        # every case?? https://invisible-island.net/ncurses/ncurses.faq.html#xterm_color
        # set encoding to None so we get bytes for consistency w/ other scrapli transports
        self.stdin, self.stdout, self.stderr = await self.session.open_session(
            term_type="xterm", encoding=None
        )

    async def _authenticate(self) -> None:
        """
        Parent method to try all means of authentication

        Args:
            N/A

        Returns:
            N/A  # noqa: DAR202

        Raises:
            ScrapliAuthenticationFailed: if authentication fails

        """
        # we already fetched host/port/user from the user input and/or the ssh config file, so we
        # want to use those explicitly. likewise we pass config file we already found. set known
        # hosts and agent to None so we can not have an agent and deal w/ known hosts ourselves
        common_args = {
            "host": self.host,
            "port": self.port,
            "username": self.auth_username,
            "known_hosts": None,
            "agent_path": None,
            "config": self.ssh_config_file,
        }

        if self.auth_private_key:
            if await self._authenticate_private_key(common_args=common_args):
                self.logger.debug(f"Authenticated to host {self.host} with public key auth")
                return
            if not self.auth_password or not self.auth_username:
                msg = (
                    f"Failed to authenticate to host {self.host} with private key "
                    f"`{self.auth_private_key}`. Unable to continue authentication, "
                    "missing username, password, or both."
                )
                self.logger.critical(msg)
                raise ScrapliAuthenticationFailed(msg)

        if not await self._authenticate_password(common_args=common_args):
            msg = f"Authentication to host {self.host} failed"
            self.logger.critical(msg)
            raise ScrapliAuthenticationFailed(msg)

        self.logger.debug(f"Authenticated to host {self.host} with password")

    async def _authenticate_private_key(self, common_args: Dict[str, Any]) -> bool:
        """
        Attempt to authenticate with key based authentication

        Args:
            common_args: Dict of kwargs that are common between asyncssh auth/open methods

        Returns:
            bool: True if authentication succeeds, otherwise False

        Raises:
            ScrapliTimeout: if authentication times out
            Exception: if unknown (i.e. not auth failed) exception occurs

        """
        try:
            self.session = await asyncio.wait_for(
                connect(
                    client_keys=self.auth_private_key, preferred_auth=("publickey",), **common_args
                ),
                timeout=self.timeout_socket,
            )
            return True
        except asyncio.TimeoutError as exc:
            msg = (
                f"Private key authentication with host {self.host} failed. "
                "Authentication Timed Out."
            )
            self.logger.exception(msg)
            raise ScrapliTimeout(msg) from exc
        except PermissionDenied:
            self.logger.critical(
                f"Private key authentication with host {self.host} failed. Authentication Error."
            )
            return False
        except Exception as exc:
            self.logger.critical(
                f"Private key authentication with host {self.host} failed. Exception: {exc}."
            )
            raise exc

    async def _authenticate_password(self, common_args: Dict[str, Any]) -> bool:
        """
        Attempt to authenticate with password/kbd-interactive authentication

        Args:
            common_args: Dict of kwargs that are common between asyncssh auth/open methods

        Returns:
            bool: True if authentication succeeds, otherwise False

        Raises:
            ScrapliTimeout: if authentication times out
            Exception: if unknown (i.e. not auth failed) exception occurs

        """
        try:
            self.session = await asyncio.wait_for(
                connect(
                    password=self.auth_password,
                    preferred_auth=(
                        "keyboard-interactive",
                        "password",
                    ),
                    **common_args,
                ),
                timeout=self.timeout_socket,
            )
            return True
        except asyncio.TimeoutError as exc:
            msg = f"Password authentication with host {self.host} failed. Authentication Timed Out."
            self.logger.exception(msg)
            raise ScrapliTimeout(msg) from exc
        except PermissionDenied:
            self.logger.critical(
                f"Password authentication with host {self.host} failed. Authentication Error."
            )
            return False
        except Exception as exc:
            self.logger.critical(
                f"Password authentication with host {self.host} failed. Exception: {exc}."
            )
            raise exc

    def _isauthenticated(self) -> bool:
        """
        Check if session is authenticated

        Args:
            N/A

        Returns:
            bool: True if authenticated, else False

        Raises:
            N/A

        """
        isauthenticated: bool = self.session._auth_complete  # pylint:  disable=W0212
        return isauthenticated

    def close(self) -> None:
        """
        Close session and socket

        Args:
            N/A

        Returns:
            N/A  # noqa: DAR202

        Raises:
            N/A

        """
        try:
            self.session.close()
        except BrokenPipeError:
            # it seems it is possible for the connection transport is_closing() to be true already
            # in some cases... since we are closing the connection anyway we will just ignore this
            # note that this seemed to only happen in github actions on ubuntu-latest w/ py3.8...
            pass
        self.session._auth_complete = False  # pylint:  disable=W0212
        self.logger.debug(f"Channel to host {self.host} closed")

    def isalive(self) -> bool:
        """
        Check if socket is alive and session is authenticated

        Args:
            N/A

        Returns:
            bool: True if socket is alive and session authenticated, else False

        Raises:
            N/A

        """
        isauthenticated: bool = self.session._auth_complete  # pylint:  disable=W0212
        # this may need to be revisited in the future, but this seems to be a good check for
        # aliveness
        try:
            if (
                isauthenticated
                and self.session._transport.is_closing() is False  # pylint:  disable=W0212
            ):
                return True
        except AttributeError:
            pass
        return False

    @requires_open_session()
    async def read(self) -> bytes:
        """
        Read data from the channel

        Args:
            N/A

        Returns:
            bytes: bytes output as read from channel

        Raises:
            ScrapliTimeout: if async read does not complete within timeout_transport interval

        """
        try:
            output: bytes = await asyncio.wait_for(
                self.stdout.read(65535), timeout=self.timeout_transport
            )
            return output
        except asyncio.TimeoutError as exc:
            msg = f"Timed out reading from transport, transport timeout: {self.timeout_transport}"
            self.logger.exception(msg)
            raise ScrapliTimeout(msg) from exc
        except ConnectionLost as exc:
            msg = "Connection lost while reading from transport"
            self.logger.exception(msg)
            raise ScrapliTimeout(msg) from exc

    @requires_open_session()
    def write(self, channel_input: str) -> None:
        """
        Write data to the channel

        Args:
            channel_input: string to send to channel

        Returns:
            N/A  # noqa: DAR202

        Raises:
            N/A

        """
        self.stdin.write(channel_input.encode())

    def set_timeout(self, timeout: int) -> None:
        """
        Set session timeout

        Args:
            timeout: timeout in seconds

        Returns:
            N/A  # noqa: DAR202

        Raises:
            N/A

        """
        self.timeout_transport = timeout
